### DISCUSSION TITLE
Does `event.listens_for` work in multiprocessing?

### DISCUSSION THREAD

**Discussion by jamesbraza** [sqlalchemy/sqlalchemy#10828] [Usage Questions] [answered]
I have one Python process with something like:

```python
from sqlalchemy import Connection, event
from sqlalchemy.orm import Mapper

from my.code.models import SomeModel

@event.listens_for(SomeModel, "after_insert")
def insert_order_to_printer(
    mapper: Mapper, connection: Connection, target: SomeModel
) -> None:
    print(target)
```

And another Python process that is performing DB inserts via SQL Alchemy.

I find the Python process that should be printing upon each DB insert never prints anything. Does `event.listens_for` work in multiprocessing?

If it helps, I am using `SQLAlchemy[postgresql_psycopgbinary]==2.0.23` with Python 3.11.7.

**Comment by CaselIT** [comment, marked as answer]
Hi,

Each process will run it own events, so you don't have a single process that receives the events of all other

**Reply by jamesbraza** [reply]
Thank you @CaselIT, appreciated. I think I would like to open a PR into the docs to add this. I am wondering, can you provide me any pointers into the source code that helps you know multiprocessing is not supported?

**Reply by CaselIT** [reply]
Well I think it would be mentioned if multiprocessing was supported since it would explain how would that work, since it's not trivial.

I don't know if a documentation add much here. @zzzeek opinions?

**Reply by zzzeek** [reply]
so....you have to do the listen in the parent process if you expect that event listener to exist in the child process, obviously if you do a listen in a child process, another child proc that does not do a listen,.....wont have that listener!  I dunno maybe a note, "if using multiprocessing, it's likely best to set the event handler in the parent procses before starting child processes".  id keep the language in the "you likely want this" category, not like "you must!" because you dont have to, its just probably what most people would want.

**Reply by jamesbraza** [reply]
Okay gotchu and I am following. I have a brief follow up question.

Lets imagine the "listener" and "inserter" are in two separate Docker containers. Is there some way to have the listener bound to the actual DB (and not a process in a Docker container), such that inserts from one Docker container can be listened to by another Docker container?

**Reply by zzzeek** [reply]
you'd have to build a message passing system between the two containers, such as broadcasting / receiving UDP packets, something like that (or grepping a log file on a common shared filesystem).   

basically the answer is, "not with SQLAlchemy's events system", which can only by itself send and receive messages in a single Python process.

**Reply by jamesbraza** [reply]
Sweet thanks!  Opened https://github.com/sqlalchemy/sqlalchemy/pull/10831 with this info. Thanks again both for your help
